File: src/main/java/li/cil/oc/util/SideTracker.java
package li.cil.oc.util;

import net.minecraftforge.forgespi.Environment;
import net.minecraftforge.fml.common.thread.EffectiveSide;

import java.util.Collections;
import java.util.Set;

public final class SideTracker {
    public static boolean isServer() {
        return Environment.get().getDist().isDedicatedServer() || EffectiveSide.get().isServer();
    }

    public static boolean isClient() {
        return !isServer();
    }
}

File: src/main/scala/li/cil/oc/util/SideTracker.scala
package li.cil.oc.util

import java.util.Collections

import cpw.mods.fml.common.FMLCommonHandler

object SideTracker {
  private val serverThreads = Collections.newSetFromMap(new java.util.WeakHashMap[Thread, java.lang.Boolean])

  def addServerThread() = serverThreads.add(Thread.currentThread())

  def isServer = FMLCommonHandler.instance.getEffectiveSide.isServer || serverThreads.contains(Thread.currentThread())

  def isClient = !isServer
}
